File: apps/workspace/server/src/lib/bootstrap.ts
import type { BootstrapPayload, WorkspaceRole } from '@echojs-ecosystem/workspace-shared'

import { db } from '../data/store'

const roleTemplates: Record<
  WorkspaceRole,
  BootstrapPayload['permissionSnapshot']['rules']
> = {
  admin: {
    dashboard: { view: true },
    user: { read: true, update: true, delete: true },
    order: { read: true, refund: true },
    catalog: { read: true, edit: true, delete: true },
    settings: { read: true, update: true },
  },
  manager: {
    dashboard: { view: true },
    user: { read: true, update: true, delete: false },
    order: { read: true, refund: true },
    catalog: { read: true, edit: true, delete: false },
    settings: { read: true, update: false },
  },
  viewer: {
    dashboard: { view: true },
    user: { read: true, update: false, delete: false },
    order: { read: true, refund: false },
    catalog: { read: true, edit: false, delete: false },
    settings: { read: true, update: false },
  },
}

export const buildBootstrap = (role: WorkspaceRole = 'manager'): BootstrapPayload => {
  const openOrders = db.orders.filter((o) => o.status !== 'refunded').length
  const activeUsers = db.users.filter((u) => u.status === 'active').length
  const revenue = db.orders.reduce((sum, o) => sum + o.total, 0)

  return {
    role,
    stats: {
      openOrders,
      activeUsers,
      revenue,
      serverTime: new Date().toISOString(),
    },
    permissionSnapshot: {
      version: 1,
      ready: true,
      rules: roleTemplates[role],
    },
  }
}

File: apps/workspace/server/internal/lib/bootstrap.go
package lib

import (
	"time"

	"github.com/echojs-ecosystem/workspace-server/internal/data"
	"github.com/echojs-ecosystem/workspace-server/internal/models"
)

var roleTemplates = map[models.WorkspaceRole]map[string]map[string]bool{
	models.RoleAdmin: {
		"dashboard": {"view": true},
		"user":      {"read": true, "update": true, "delete": true},
		"order":     {"read": true, "refund": true},
		"catalog":   {"read": true, "edit": true, "delete": true},
		"settings":  {"read": true, "update": true},
	},
	models.RoleManager: {
		"dashboard": {"view": true},
		"user":      {"read": true, "update": true, "delete": false},
		"order":     {"read": true, "refund": true},
		"catalog":   {"read": true, "edit": true, "delete": false},
		"settings":  {"read": true, "update": false},
	},
	models.RoleViewer: {
		"dashboard": {"view": true},
		"user":      {"read": true, "update": false, "delete": false},
		"order":     {"read": true, "refund": false},
		"catalog":   {"read": true, "edit": false, "delete": false},
		"settings":  {"read": true, "update": false},
	},
}

func BuildBootstrap(store *data.Store, role models.WorkspaceRole) models.BootstrapPayload {
	if role == "" {
		role = models.RoleManager
	}

	orders, _ := store.AllOrders()
	users, _ := store.AllUsers()

	openOrders := 0
	for _, order := range orders {
		if order.Status != models.OrderRefunded {
			openOrders++
		}
	}

	activeUsers := 0
	for _, user := range users {
		if user.Status == models.StatusActive {
			activeUsers++
		}
	}

	var revenue float64
	for _, order := range orders {
		revenue += order.Total
	}

	rules, ok := roleTemplates[role]
	if !ok {
		rules = roleTemplates[models.RoleManager]
	}

	payload := models.BootstrapPayload{
		Role: role,
		Stats: models.DashboardStats{
			OpenOrders:  openOrders,
			ActiveUsers: activeUsers,
			Revenue:     revenue,
			ServerTime:  time.Now().UTC().Format(time.RFC3339Nano),
		},
	}
	payload.PermissionSnapshot.Version = 1
	payload.PermissionSnapshot.Ready = true
	payload.PermissionSnapshot.Rules = rules

	return payload
}
